### PR TITLE
Don't retain device handle references in sanitizer layer.

### PR DESCRIPTION
**Pull request by aarongreig**
Also use uint32_t for Debug value.

### CODE REVIEW

**Comment by aarongreig** [comment]
This fixes https://github.com/intel/llvm/issues/14658 and all the other asan fails.

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 1]
```diff
@@ -708,8 +708,7 @@ ur_result_t SanitizerInterceptor::prepareLaunch(
         };
 
         // Write debug
-        // We use "uint64_t" here because EnqueueWriteGlobal will fail when it's "uint32_t"
```
Comment (aarongreig) [inline source/loader/layers/sanitizer/asan_interceptor.cpp:711]: I didn't observe this fail when running locally, and I discovered that changing this to `uint32_t` fixes a test for the Debug option. Was this a fix for a specific platform @AllanZyne @yingcong-wu?

**Comment by aarongreig** [comment]
LLVM PR https://github.com/intel/llvm/pull/14719

File: source/loader/layers/sanitizer/asan_interceptor.cpp (hunk 1 above)
Comment (yingcong-wu) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:711]: A fix related to this var is purposed here: https://github.com/oneapi-src/unified-runtime/pull/1878

**Comment by yingcong-wu** [comment]
Could you please explain why is this change needed?

File: source/loader/layers/sanitizer/asan_interceptor.cpp (hunk 1 above)
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:711]: > I didn't observe this fail when running locally, and I discovered that changing this to `uint32_t` fixes a test for the Debug option. Was this a fix for a specific platform @AllanZyne @yingcong-wu?

I remember correctly, `uint32_t` was not worked on PVC. 
If you tested this and it worked, I also think it'd be better to use `uint32_t`.

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 2]
```diff
@@ -708,8 +708,7 @@ ur_result_t SanitizerInterceptor::prepareLaunch(
...
-        uint64_t Debug = Options(logger).Debug ? 1 : 0;
+        uint32_t Debug = Options(logger).Debug ? 1 : 0;
         EnqueueWriteGlobal(kSPIR_AsanDebug, &Debug, sizeof(Debug));
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.cpp:712]: I think we can use Options directly, no need to use a temporary variable
```suggestion
        EnqueueWriteGlobal(kSPIR_AsanDebug, &Option(logger).Debug, sizeof(Option(logger).Debug));
```

Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:712]: If the type of `Option::Debug` is not correct, we can change it at `Option`

**Comment by AllanZyne** [comment]
Could you please explain why this change is needed?

File: source/loader/layers/sanitizer/asan_interceptor.cpp (hunk 1 above)
Comment (yingcong-wu) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:711]: PR #1878 is merged recently, you can see if that fix the problem you are trying to fix with this change.

**Comment by aarongreig** [comment]
> Could you please explain why not retain device handle references in sanitizer layer?

Following the PI->UR port we started seeing fails in a handful of AddressSanitizer e2e tests that were kind of similar to the issue https://github.com/oneapi-src/unified-runtime/pull/1879 is trying to solve: the sycl RT unloads the adapter and tears down the loader before the `SanitizerInterceptor` destructor is called, so by the time device references were released [here](https://github.com/oneapi-src/unified-runtime/blob/907f1542c7498b2115c3eff8286dbd09e50c9457/source/loader/layers/sanitizer/asan_interceptor.hpp#L56) the adapter was no longer around to properly release the device. My initial approach to this was to add reference counting for all the devices saved in [m_DeviceMap](https://github.com/oneapi-src/unified-runtime/blob/907f1542c7498b2115c3eff8286dbd09e50c9457/source/loader/layers/sanitizer/asan_interceptor.hpp#L248) but that quickly got complicated. I figured that since the sanitizer layer only uses devices while the application still considers them valid (e.g. during [memory allocations](https://github.com/oneapi-src/unified-runtime/blob/907f1542c7498b2115c3eff8286dbd09e50c9457/source/loader/layers/sanitizer/asan_interceptor.cpp#L177), [kernel launches](https://github.com/oneapi-src/unified-runtime/blob/907f1542c7498b2115c3eff8286dbd09e50c9457/source/loader/layers/sanitizer/asan_interceptor.cpp#L337) etc.) we can get away without artificially extending the lifetime of these handles and solve the SYCL RT/UR destruction order problems in a simpler fashion.

**Comment by aarongreig** [comment]
it's possible that https://github.com/oneapi-src/unified-runtime/pull/1879 could also solve this, although in its current state that PR actually causes all the AddressSanitizer e2e tests to fail on the PI2UR-main branch

**Comment by pbalcer** [comment]
> the sycl RT unloads the adapter and tears down the loader before the SanitizerInterceptor destructor is called

This should not be happening. The layer objects are teardown'ed inside of [`urLoaderTeardown`](https://github.com/oneapi-src/unified-runtime/blob/main/source/loader/ur_lib.cpp#L210), prior to the loader's own context (which includes handle maps). Unless the sanitizer state is stored somewhere in a global variable. In which case the fix is to get rid of the global state.

The loader does have one hack that causes `getContext()` to actually re-initialize all the state if its called after the loader has been already destroyed. This is to prevent misbehaving applications from crashing, but leaking data is expected in such cases. But in that case, the SanitizerInterceptor should never have been called.

**Review by pbalcer — APPROVED** [review]
I think I get it now. The adapter is destroyed before all device handles are released...

We need to start running tests with full validation layer enabled. It would have caught it.

**Comment by aarongreig** [comment]
yeah I double checked and it's literally just that these release calls happen when the layers are (correctly) torn down in `urLoaderTearDown`, which necessarily happens after the `urAdapterRelease` call renders them invalid

**Comment by aarongreig** [comment]
the multi device fail is unrelated

**Review by AllanZyne — APPROVED** [review]
I think this fix can be treated as a hotfix, but not final solution.
Can you add a comment in "struct DeviceInfo" to remind me about this?
Thanks.

**Comment by aarongreig** [comment]
> I made an issue for this, so please add
> 
> "Fixes: [intel/llvm#14825](https://github.com/intel/llvm/issues/14825)" to the commit description, thanks

I've added it here https://github.com/intel/llvm/pull/14719 because that change will re-enable the tests and actually land the fix

**Comment by AllanZyne** [comment]
Hi @pbalcer, please merge this PR. 
This bugfix is needed for 2025.0.

**Review by zhaomaosu — APPROVED** [review]
lgtm

**Pull request merged**